src/hivpy/hiv_testing.py

```diff
@@ -0,0 +1,145 @@
+import importlib.resources
+import operator as op
+from datetime import timedelta
+
+import hivpy.column_names as col
+
+from .common import rng
+from .hiv_testing_data import HIVTestingData
+
+
+class HIVTestingModule:
+
+    def __init__(self, **kwargs):
+
+        # init hiv testing data
+        with importlib.resources.path("hivpy.data", "hiv_testing.yaml") as data_path:
+            self.ht_data = HIVTestingData(data_path)
+
+        self.date_start_testing = self.ht_data.date_start_testing
+        self.init_rate_first_test = self.ht_data.init_rate_first_test
+        self.eff_max_freq_testing = self.ht_data.eff_max_freq_testing
+        self.test_scenario = self.ht_data.test_scenario
+        self.no_test_if_np0 = self.ht_data.no_test_if_np0
+        self.test_targeting = self.ht_data.test_targeting.sample()
+        self.date_test_rate_plateau = self.ht_data.date_test_rate_plateau.sample()
+        self.an_lin_incr_test = self.ht_data.an_lin_incr_test.sample()
+
+        # eff_max_freq_testing is used as an index to pick the correct
+        # minimum number of days to wait between tests from this list
+        self.days_to_wait = [365, 180, 90]  # 12 months, 6 months, 3 months
+
+        self.rate_first_test = 0
+        self.rate_rep_test = 0
+        # TODO: meant to be a personal variable but currently always just set to test_targeting
+        self.eff_test_targeting = self.test_targeting
+        # scenario has a high chance to change eff_test_targeting
+        if self.test_scenario == 1:
+            r = rng.uniform()
+            if r < 0.45:
+                self.eff_test_targeting = 2
+            elif r < 0.9:
+                self.eff_test_targeting = 5
+        # FIXME: move this to a yaml later
+        self.covid_disrup_affected = False
+        self.testing_disrup_covid = False
+
+    def update_hiv_testing(self, pop):
+        """
+        Update which individuals in the population have been tested.
+        COVID disruption is factored in.
+        """
+        # testing occurs after a certain year if there is no covid disruption
+        if ((pop.date.year >= (self.date_start_testing + 5.5))
+           & ((not self.covid_disrup_affected) | (not self.testing_disrup_covid))):
```

Comment:
I think there needs to be a change here, but we should ask the HIV team about this. 

- In the SAS code `date_start_testing` is 2003.5, and so `date_start_testing + 5.5` is 2009 
- `pop.date.year` is an int, which is okay if we're comparing to 2009, but if this value were not a whole number then we would be in trouble. It woule be better to calculate this as an actual date and compare rather than using the same calculation as in the SAS code, because they represent their date as a decimal number like 2003.5 for the middle of 2003
- This calculation also just strikes me as suspicious; would we still be adding 5.5 if `date_start_testing` were not 2003.5? Is there a significance to 5.5 years? I think we should check up on this.

Reply:
In general we also have the potential issue of an int >= a float may fail on the equal case; should be fine in this case since exactly 0.5 is representable in FP but something we should avoid in general I think when the exact case is important.

Reply:
`date_start_testing + 5.5` also comes up several times in the following code so might be worth giving it its own variable. But again we can address that after we've found out a bit more about what it means